src/variants/abalone/directions.ts

```diff
@@ -0,0 +1,165 @@
+};
+
+export const candidateLineDirs = (origToDestDirection: DiagonalDirectionString): DirectionString[] => {
+  const directionMap: { [key in DiagonalDirectionString]: DirectionString[] } = {
+    [DiagonalDirectionString.UpLeft]: [HorizontalDirectionString.Left, DiagonalDirectionString.UpLeft],
+    [DiagonalDirectionString.UpRight]: [
+      DiagonalDirectionString.UpLeft,
+      DiagonalDirectionString.UpRight,
+      HorizontalDirectionString.Right,
+    ],
+    [DiagonalDirectionString.DownRight]: [HorizontalDirectionString.Right, DiagonalDirectionString.DownRight],
+    [DiagonalDirectionString.DownLeft]: [
+      DiagonalDirectionString.DownRight,
+      DiagonalDirectionString.DownLeft,
+      HorizontalDirectionString.Left,
+    ],
+  };
+  return directionMap[origToDestDirection];
+};
+
+export const deducePotentialSideDirs = (
+  origToDestDirection: DiagonalDirectionString,
+  lineDirection: DirectionString,
+): DirectionString[] => {
+  switch (origToDestDirection) {
+    case DiagonalDirectionString.DownLeft:
+      switch (lineDirection) {
+        case DiagonalDirectionString.DownLeft:
+          return [HorizontalDirectionString.Left, DiagonalDirectionString.DownRight];
+        case DiagonalDirectionString.DownRight:
+          return [DiagonalDirectionString.DownLeft];
+        case HorizontalDirectionString.Left:
+          return [DiagonalDirectionString.DownLeft];
+        default:
+          return [];
+      }
+    case DiagonalDirectionString.UpRight:
+      switch (lineDirection) {
+        case DiagonalDirectionString.UpLeft:
+          return [DiagonalDirectionString.UpRight];
+        case DiagonalDirectionString.UpRight:
+          return [HorizontalDirectionString.Right, DiagonalDirectionString.UpLeft];
+        case HorizontalDirectionString.Right:
+          return [DiagonalDirectionString.UpRight];
+        default:
+          return [];
+      }
+    case DiagonalDirectionString.UpLeft:
+      switch (lineDirection) {
+        case DiagonalDirectionString.UpLeft:
+          return [HorizontalDirectionString.Left];
+        case HorizontalDirectionString.Left:
+          return [DiagonalDirectionString.UpLeft];
+        default:
+          return [];
+      }
+    case DiagonalDirectionString.DownRight:
+      switch (lineDirection) {
+        case DiagonalDirectionString.DownRight:
+          return [HorizontalDirectionString.Right];
+        case HorizontalDirectionString.Right:
+          return [DiagonalDirectionString.DownRight];
+        default:
+          return [];
+      }
+    default:
+      return [];
+  }
+};
+
+const traverseUntil = (pos: cg.Key, stop: (pos: cg.Key) => boolean, direction: DirectionString): cg.Key[] => {
+  const nextPos = move(pos, direction);
+  if (nextPos) {
+    const rest = stop(nextPos) ? [] : traverseUntil(nextPos, stop, direction);
+    return [nextPos, ...rest];
+  } else {
+    return [];
+  }
+};
+
+const isValidKey = (key: cg.Key): boolean => {
```

Comment:
using regex would be simpler here?
`/^(a[1-5]|b[1-6]|c[1-7]|d[1-8]|e[1-9]|f[2-9]|g[3-9]|h[4-9]|i[5-9])$/`

Reply:
yes indeed !
thank you